ospf6d/ospf6_auth_trailer.c

```diff
@@ -0,0 +1,817 @@
+		     (oh_seqnum_l > on->seqnum_l)) {
+			/* valid sequence number received */
+			on->seqnum_h = oh_seqnum_h;
+			on->seqnum_l = oh_seqnum_l;
+		} else {
+			if (IS_OSPF6_DEBUG_AUTH_RX) {
+				zlog_warn("RECV[%s] : Nbr(%s) Auth Sequence number mismatch",
+					  oi->interface->name, on->name);
+				zlog_warn("nbr_seq_l %u, nbr_seq_h %u, hdr_seq_l %u, hdr_seq_h %u",
+					  on->seqnum_l, on->seqnum_h,
+					  oh_seqnum_l, oh_seqnum_h);
+			}
+
+			oi->at_data.rx_drop++;
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+/* Starting point of packet process function. */
+int ospf6_auth_check_digest(struct ospf6_header *oh, struct ospf6_interface *oi,
+			    struct in6_addr *src)
+{
+	uint32_t hash_len = KEYCHAIN_MAX_HASH_SIZE;
+	unsigned char apad[hash_len];
+	unsigned char temp_hash[hash_len];
+	struct ospf6_auth_hdr *ospf6_auth;
+	uint32_t ipv6_addr_size = sizeof(struct in6_addr);
+	struct keychain *keychain = NULL;
+	struct key *key = NULL;
+	char *auth_str = NULL;
+	uint16_t auth_len = 0;
+	uint8_t hash_algo = 0;
+	uint16_t oh_len = ntohs(oh->length);
+
+	if (oi->at_data.flags == 0)
+		return -2;
+
+	ospf6_auth = (struct ospf6_auth_hdr *)((uint8_t *)oh + oh_len);
+	if (CHECK_FLAG(oi->at_data.flags, OSPF6_AUTH_TRAILER_KEYCHAIN)) {
+		keychain = keychain_lookup(oi->at_data.keychain);
+		if (!keychain) {
+			if (IS_OSPF6_DEBUG_AUTH_RX)
+				zlog_warn("RECV[%s]: Keychain does't exist for %s",
+					  oi->interface->name,
+					  ospf6_message_type(oh->type));
+			oi->at_data.rx_drop++;
+			return -1;
+		}
+
+		key = key_lookup_for_accept(keychain, ntohs(ospf6_auth->id));
+		if (!key) {
+			if (IS_OSPF6_DEBUG_AUTH_RX)
+				zlog_warn("RECV[%s]: Auth, Invalid SA for %s",
+					  oi->interface->name,
+					  ospf6_message_type(oh->type));
+			oi->at_data.rx_drop++;
+			return -1;
+		}
+
+		if (key && key->string
+		    && key->hash_algo != KEYCHAIN_ALGO_NULL) {
+			auth_str = key->string;
+			hash_algo = key->hash_algo;
+		} else {
+			if (IS_OSPF6_DEBUG_AUTH_RX)
+				zlog_warn("RECV[%s]: Incomplete keychain config for %s",
+					  oi->interface->name,
+					  ospf6_message_type(oh->type));
+			oi->at_data.rx_drop++;
+			return -1;
+		}
+	} else if (CHECK_FLAG(oi->at_data.flags,
+			      OSPF6_AUTH_TRAILER_MANUAL_KEY)) {
+		auth_str = oi->at_data.auth_key;
+		hash_algo = oi->at_data.hash_algo;
+	}
+
+	if (!auth_str)
+		return -2;
+
+	hash_len = keychain_get_hash_len(hash_algo);
+	memset(apad, 0, sizeof(hash_len));
+	memset(temp_hash, 0, sizeof(hash_len));
+
+	/* start digest verification */
+	memcpy(apad, src, ipv6_addr_size);
+	memcpy(apad + ipv6_addr_size, ospf6_hash_apad_max,
+	       (hash_len - ipv6_addr_size));
+
+	auth_len = ntohs(ospf6_auth->length);
+
+	memcpy(temp_hash, ospf6_auth->data, hash_len);
+	memcpy(ospf6_auth->data, apad, hash_len);
+
+	ospf6_auth_update_digest(oi, oh, ospf6_auth, auth_str, auth_len,
+				 (oh_len + auth_len), hash_algo);
+
+	return !(memcmp(temp_hash, ospf6_auth->data, hash_len));
```

Comment:
This should use `CRYPTO_memcmp` if OpenSSL is in use, to reduce the risk of side channel attacks.  (Without OpenSSL I don't think we have a constant-time memcmp available.)

( https://www.openssl.org/docs/man1.1.1/man3/CRYPTO_memcmp.html )

Reply:
Do you want me to add #defs for memcmp with openssl and without openssl ?

Reply:
done.